src/heartratetask/HeartRateTask.cpp

```diff
@@ -110,10 +92,114 @@ void HeartRateTask::StartMeasurement() {
   heartRateSensor.Enable();
   ppg.Reset(true);
   vTaskDelay(100);
+  measurementStart = xTaskGetTickCount();
 }
 
 void HeartRateTask::StopMeasurement() {
   heartRateSensor.Disable();
   ppg.Reset(true);
   vTaskDelay(100);
 }
+
+void HeartRateTask::StartWaiting() {
+  StopMeasurement();
+  backgroundWaitingStart = xTaskGetTickCount();
+}
+
+void HeartRateTask::HandleBackgroundWaiting() {
+  if (!IsBackgroundMeasurementActivated()) {
+    return;
+  }
+
+  TickType_t ticksSinceWaitingStart = xTaskGetTickCount() - backgroundWaitingStart;
+  if (ticksSinceWaitingStart >= GetHeartRateBackgroundMeasurementIntervalInTicks()) {
+    state = States::BackgroundMeasuring;
+    StartMeasurement();
+  }
+}
+
+void HeartRateTask::HandleSensorData(int* lastBpm) {
+  int8_t ambient = ppg.Preprocess(heartRateSensor.ReadHrs(), heartRateSensor.ReadAls());
+  int bpm = ppg.HeartRate();
+
+  // If ambient light detected or a reset requested (bpm < 0)
+  if (ambient > 0) {
+    // Reset all DAQ buffers
+    ppg.Reset(true);
+  } else if (bpm < 0) {
+    // Reset all DAQ buffers except HRS buffer
+    ppg.Reset(false);
+    // Set HR to zero and update
+    bpm = 0;
+  }
+
+  if (*lastBpm == 0 && bpm == 0) {
+    controller.Update(Controllers::HeartRateController::States::NotEnoughData, bpm);
+  }
+
+  if (bpm != 0) {
+    *lastBpm = bpm;
+    controller.Update(Controllers::HeartRateController::States::Running, bpm);
+    if (state == States::Measuring || IsContinuosModeActivated()) {
+      return;
+    }
+    if (state == States::BackgroundMeasuring) {
+      state = States::BackgroundWaiting;
+      StartWaiting();
+    }
+  }
+  TickType_t ticksSinceMeasurementStart = xTaskGetTickCount() - measurementStart;
+  if (bpm == 0 && state == States::BackgroundMeasuring && !IsContinuosModeActivated() &&
+      ticksSinceMeasurementStart >= DURATION_UNTIL_BACKGROUND_MEASURMENT_IS_STOPPED) {
+    state = States::BackgroundWaiting;
+    StartWaiting();
+  }
+  if (bpm == 0 && state == States::BackgroundMeasuring &&
+      xTaskGetTickCount() - measurementStart >= DURATION_UNTIL_BACKGROUND_MEASURMENT_IS_STOPPED) {
+    state = States::BackgroundWaiting;
+    StartWaiting();
+  }
```

Comment:
The second `if` block looks almost like the first one, except that it is not handling `IsContinuousModeActivated`.. so if I am reading this correct (and to be honest, this function is very hard for me to grasp in what it is doing when and why), the `IsContinousModeActivated` isn't handled properly so it will still wait in theory. Why it still might work is, that the `waiting` part is getting a time-to-wait of `0` in `GetHeartRateBackgroundMeasurementIntervalInTicks` as the continous setting should fall into the default value.

But if I am reading it correct I would assume that this code should be changed like that:

```suggestion
  TickType_t ticksSinceMeasurementStart = xTaskGetTickCount() - measurementStart;
  if (bpm == 0 && state == States::BackgroundMeasuring && !IsContinuosModeActivated() &&
      ticksSinceMeasurementStart >= DURATION_UNTIL_BACKGROUND_MEASURMENT_IS_STOPPED) {
    state = States::BackgroundWaiting;
    StartWaiting();
  }
```

Reply:
Besides that, I would like to test this out on my PineTime (sealed) but even though the "upload" of the dfu file succeeded, it still boots to the original one.. independently if I build version 1.14.0 or (to have a newer one) 1.14.1 as a test.

Not sure what I do wrong.

Reply:
> But if I am reading it correct I would assume that this code should be changed like that:

I think that the duplication happened during one of the many rebases and merges I already had to do to keep the PR up to date. Your code suggestion is correct

> ... "upload" of the dfu file succeeded, it still boots to the original one.. ...

This behavior is expected if there is a problem with booting a new version (like something immediately throwing an error on boot). In these cases the watch reverts back to the old version so that it doesn't get bricked and unusable.

I also had this problem sometimes because of different reasons, but I thought that it would have been fixed, since I'm running this PR on my sealed PineTime for nearly as long as the PR is open.
I'll fix the issues mentioned already and see if the fixes it, otherwise I'll try to find out what could cause this in the current version.

Reply:
thanks! I will try that later today!

Reply:
The bootloader doesn't care what the firmware version number is. The only thing it does in that regard is store a validation bit to see if it should rollback or not.